doc/pmon/sonic_thermal_control_test_plan.md

```diff
@@ -0,0 +1,250 @@
+
+#### Invalid value policy file
+
+In invalid_value_policy.json, the file content contains minus value of target speed. We couldn't cover all invalid value here because there are too many possibilities. The major purpose of this configuration file is to verify thermal control daemon would not crash while loading such an invalid configuration files. For other negative test cases, they are already covered by unit test. The invalid_value_policy.json content is like:
+
+```json
+{
+    "info_types": [
+        {
+            "type": "fan_info"
+        },
+        {
+            "type": "psu_info"
+        }
+    ],
+    "policies": [
+        {
+            "name": "any PSU absence",
+            "conditions": [
+                {
+                    "type": "fan.any.absence"
+                }
+            ],
+            "actions": [
+                {
+                    "type": "fan.all.set_speed",
+                    "speed": "-1"
+                }
+            ]
+        }
+    ]
+}
+```
+
+## Test Cases
+
+### Show FAN Status Test
+
+Show FAN status test verifies that all FAN related information can be shown correctly via `show platform fanstatus`.
+
+#### Procedure
+
+1. Testbed setup.
+2. Unlink FAN related sysfs and fill mock data for "presence", "speed", "status", "target_speed", "led status".
+3. Issue command `show platform fanstatus`.
+4. Record the command output.
+5. Verify that command output matches the mock data.
+6. Restore FAN related soft links in sysfs.
+
+### Show Thermal Status Test
+
+Show thermal status test verifies that all thermal related information can be shown correctly via `show platform temperature`.
+
+#### Procedure
+
+1. Testbed setup.
+2. Unlink thermal related sysfs and fill mock data for "temperature", "high_threshold", "high_critical_threshold".
+3. Issue command `show platform temperature`.
+4. Record the command output.
+5. Verify that command output matches the mock data.
+6. Restore FAN related soft links in sysfs.
+
+### FAN Test
+
+FAN test verifies that proper action should be taken for conditions including: FAN absence, FAN over speed, FAN under speed.
+
+#### Procedure
+
+1. Testbed setup.
+2. Copy valid_policy.json to pmon docker and backup the original one.
+3. Restart pmon service by command `systemctl restart pmon` to trigger thermal control daemon reload policy configuration file.
+4. Unlink FAN related sysfs and make mock data: first FAN absence.
+5. Wait for at least 65 seconds. Verify target speed of all FANs are set to 100% according to valid_policy.json. Verify there is a warning log for FAN absence.
+6. Make mock data: first FAN presence.
+7. Wait for at least 65 seconds. Verify target speed of all FANs are set to 60% according to valid_policy.json. Verify there is a notice log for FAN presence.
+8. Make mock data: first FAN speed exceed threshold(speed < target speed), second FAN speed exceed threshold(speed > target speed).
```

Comment:
how can you create moc data if thermal control is enabled and running in the system regardless of the moc?

Reply:
Currently device status such as FAN speed, FAN presence status are read from sysfs by platform API. Those sysfs files are soft links to some "real files" controlled by hardware manager. So Stephen suggests to unlink those sysfs files so that device status will not be updated by hardware manager anymore, and we can mock data after that.

Reply:
@Junchao-Mellanox The mocking procedure is very specific to mellanox platform. Not sure this will work for others.